testctrl/svc/orch/executor_test.go

```diff
@@ -13,105 +15,166 @@ import (
 )
 
 func TestKubeExecutorProvision(t *testing.T) {
-	// test provision with successful pods
-	fakePodInf := newFakePodInterface(t)
-
-	driver := types.NewComponent(testContainerImage, types.DriverComponent)
-	server := types.NewComponent(testContainerImage, types.ServerComponent)
-	client := types.NewComponent(testContainerImage, types.ClientComponent)
-
-	components := []*types.Component{server, client, driver}
-	session := types.NewSession(driver, components[:2], nil)
-
-	e := newKubeExecutor(0, fakePodInf, nil, nil)
-	eventChan := make(chan *PodWatchEvent)
-	e.eventChan = eventChan
-	e.session = session
-
-	go func() {
-		for _, c := range components {
-			eventChan <- &PodWatchEvent{
-				SessionName:   session.Name,
-				ComponentName: c.Name,
-				Pod:           nil,
-				PodIP:         "127.0.0.1",
-				Health:        Ready,
-				Error:         nil,
-			}
-		}
-	}()
 
-	if err := e.provision(); err != nil {
-		t.Fatalf("unexpected error in provision: %v", err)
+	cases := []struct {
+		description string
+		events      []fakeWatchEvent
+		ctxTimeout  time.Duration
+		errors      bool
+	}{
+		{
+			description: "successful pods",
+			events: []fakeWatchEvent{
+				{
+					component: types.ServerComponent,
+					health:    Ready,
+					podIP:     "127.0.0.1",
+				},
+				{
+					component: types.ClientComponent,
+					health:    Ready,
+					podIP:     "127.0.0.1",
+				},
+				{
+					component: types.DriverComponent,
+					health:    Ready,
+					podIP:     "127.0.0.1",
+				},
+			},
+			errors: false,
+		},
+		{
+			description: "failed driver pod",
+			events: []fakeWatchEvent{
+				{
+					component: types.ServerComponent,
+					health:    Ready,
+					podIP:     "127.0.0.1",
+				},
+				{
+					component: types.ClientComponent,
+					health:    Ready,
+					podIP:     "127.0.0.1",
+				},
+				{
+					component: types.DriverComponent,
+					health:    Failed,
+					podIP:     "127.0.0.1",
+				},
+			},
+			errors: true,
+		},
+		{
+			description: "cancelled context",
+			ctxTimeout:  1 * time.Millisecond * timeMultiplier,
+			events: []fakeWatchEvent{
+				{
+					component: types.ServerComponent,
+					sleep:     10 * time.Second * timeMultiplier,
+					health:    Failed,
+					podIP:     "127.0.0.1",
+				},
```

Comment:
NTS: It may be a more thorough test to include the other components. In case validations cause a failure in the future, like missing a driver or client.